editor/src/utils/react-performance.ts

```diff
@@ -209,141 +209,188 @@ function failSafeMemoEqualityFunction(componentDisplayName: string, severity: 's
-  // I've made it return oldValue or newValue instead of true and false
-  // and the recursion only stops if we get to return oldValue
-  // otherwise we still drill into non-equal objects to try and find
-  // keys that can be made equal
-
-  // If we're more than 100 frames deep, let's just call it a day
+  valueStackSoFar: Set<any>,
+) {
   if (stackSizeInner > 100) {
-    return possibleNewValueInner
-  }
-
-  if (oldValueInner == null) {
-    return possibleNewValueInner
+    return possibleNewValue
   }
-  var isArray = Array.isArray
-  var keyList = Object.keys
-  var hasProp = Object.prototype.hasOwnProperty
-
-  if (oldValueInner === possibleNewValueInner) return oldValueInner
 
   // We appear to have looped back on ourselves,
   // escape by just returning the value.
-  if (valueStackSoFar.includes(possibleNewValueInner)) {
-    return possibleNewValueInner
+  if (valueStackSoFar.has(possibleNewValue)) {
+    return possibleNewValue
   }
+  // mutation
+  valueStackSoFar.add(possibleNewValue)
 
-  const newValueStack: Array<any> = [...valueStackSoFar, possibleNewValueInner]
+  if (oldValue === possibleNewValue) return oldValue
 
   if (
-    oldValueInner &&
-    possibleNewValueInner &&
-    typeof oldValueInner == 'object' &&
-    typeof possibleNewValueInner == 'object'
+    oldValue &&
+    possibleNewValue &&
+    typeof oldValue == 'object' &&
+    typeof possibleNewValue == 'object'
   ) {
-    const arrA = isArray(oldValueInner)
-    const arrB = isArray(possibleNewValueInner)
-
-    // reusable vars for all the performance
-    let i: number = 0
-    let length: number = 0
-    let key: string = ''
+    if (oldValue.constructor !== possibleNewValue.constructor) return possibleNewValue
 
-    if (arrA && arrB) {
-      length = possibleNewValueInner.length
+    var length, i, keys
+    if (Array.isArray(oldValue)) {
+      let newArrayToReturn: any[] = []
+      let canSaveOldArray = true
 
-      var newArrayToReturn: any[] = []
-      var canSaveOldArray = true
-      if (length != oldValueInner.length) {
+      length = possibleNewValue.length
+      if (length != oldValue.length) {
         canSaveOldArray = false
       }
       for (i = length; i-- !== 0; ) {
         // try to recurse into the array item here and save it if possible
         newArrayToReturn[i] = keepDeepReferenceEqualityInner(
-          oldValueInner[i],
-          possibleNewValueInner[i],
+          oldValue[i],
+          possibleNewValue[i],
           stackSizeInner + 1,
-          newValueStack,
+          valueStackSoFar,
         )
-        if (oldValueInner[i] !== newArrayToReturn[i]) {
+        if (oldValue[i] !== newArrayToReturn[i]) {
           canSaveOldArray = false
         }
       }
       if (canSaveOldArray) {
-        return oldValueInner
+        return oldValue
       } else {
         return newArrayToReturn
       }
     }
 
-    if (arrA != arrB) return possibleNewValueInner
-
-    var dateA = oldValueInner instanceof Date,
```

Comment:
Since this has gone, does that mean values of type `Date` are handled by the regular object handling now?

Reply:
+1 to this question

Reply:
I'll write a test case

Reply:
I added two unit tests to test Dates. everything seems to work correctly!